python/cuda_parallel/cuda/parallel/experimental/_cccl.py

```diff
@@ -0,0 +1,209 @@
+        types.float64: TypeEnum.FLOAT64,
+    }
+    if numba_type in mapping:
+        return mapping[numba_type]
+    return TypeEnum.STORAGE
+
+
+# TODO: replace with functools.cache once our docs build environment
+# is upgraded to at least Python 3.9
+@functools.lru_cache(maxsize=None)
+def _numba_type_to_info(numba_type):
+    context = cuda.descriptor.cuda_target.target_context
+    value_type = context.get_value_type(numba_type)
+    size = value_type.get_abi_size(context.target_data)
+    alignment = value_type.get_abi_alignment(context.target_data)
+    return TypeInfo(size, alignment, _type_to_enum(numba_type))
+
+
+@functools.lru_cache(maxsize=None)
+def _numpy_type_to_info(numpy_type):
+    numba_type = numba.from_dtype(numpy_type)
+    return _numba_type_to_info(numba_type)
+
+
+def _device_array_to_cccl_iter(array):
+    dtype = array.dtype
+    info = _numpy_type_to_info(dtype)
+    # Note: this is slightly slower, but supports all ndarray-like objects as long as they support CAI
+    # TODO: switch to use gpumemoryview once it's ready
+    return Iterator(
+        info.size,
+        info.alignment,
+        IteratorKind.POINTER,
+        Op(),
+        Op(),
+        info,
+        array.__cuda_array_interface__["data"][0],
+    )
+
+
+def _iterator_to_cccl_iter(it):
+    context = cuda.descriptor.cuda_target.target_context
+    numba_type = it.numba_type
+    size = context.get_value_type(numba_type).get_abi_size(context.target_data)
+    alignment = context.get_value_type(numba_type).get_abi_alignment(
+        context.target_data
+    )
+    (advance_abi_name, advance_ltoir), (deref_abi_name, deref_ltoir) = it.ltoirs.items()
+    advance_op = Op(
+        OpKind.STATELESS,
+        advance_abi_name.encode("utf-8"),
+        ctypes.c_char_p(advance_ltoir),
+        len(advance_ltoir),
+        1,
+        1,
+        None,
+    )
+    deref_op = Op(
+        OpKind.STATELESS,
+        deref_abi_name.encode("utf-8"),
+        ctypes.c_char_p(deref_ltoir),
+        len(deref_ltoir),
+        1,
+        1,
+        None,
+    )
+    return Iterator(
+        size,
+        alignment,
+        OpKind.STATEFUL,
+        advance_op,
+        deref_op,
+        _numba_type_to_info(it.value_type),
+        it.state,
+    )
+
+
+def type_enum_as_name(enum_value):
+    assert isinstance(enum_value, int)
+    return (
+        "int8",
+        "int16",
+        "int32",
+        "int64",
+        "uint8",
+        "uint16",
+        "uint32",
+        "uint64",
+        "float32",
+        "float64",
+        "STORAGE",
+    )[enum_value]
+
+
+def to_cccl_iter(array_or_iterator):
+    from cuda.parallel.experimental.iterators._iterators import IteratorBase
+
+    if isinstance(array_or_iterator, IteratorBase):
+        return _iterator_to_cccl_iter(array_or_iterator)
+    return _device_array_to_cccl_iter(array_or_iterator)
+
+
+def host_array_to_value(array):
+    dtype = array.dtype
+    info = _numpy_type_to_info(dtype)
```

Comment:
Inline `dtype` here? (I think I suggested that before, but not sure if you missed it or like it better this way.)

Reply:
Done - I also recall you suggesting it before. Apologies for missing that!